reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,399 @@
+## Summary
+### General Motivation
+
+The current node affinity feature is relatively simple, it can only affinity with one node. 
+Many general functions cannot be supported.
+
+Current node affinity feature don't support functions:
+1. Don't support affinity to a batch of nodes. Current only support affinity to one node.
+2. Don't support anti-affinity to a node of a batch of nodes. 
+3. Don't support to classify nodes by labels, and then affinity or anti-affinity to a certain type of nodes.
+4. Don't support affinity/anti-affinity to nodes by node ip or node name. Current only support node id identification.
+
+This REP is to solve the above problems.
+
+
+### Should this change be within `ray` or outside?
+
+Yes, this will be a complement to ray core's ability to flexibly schedule actors/tasks/node.
+
+## Stewardship
+### Required Reviewers
+ 
+@ericl @stephanie-wang @wumuzi520 SenlinZhu @Chong Li  @scv119 (Chen Shen) @Sang Cho @jjyao (Jiajun Yao) @Yi Cheng
+### Shepherd of the Proposal (should be a senior committer)
+
+
+## Design and Architecture
+
+### Brief idea
+1. Introduct the concept of the static labels of nodes. Add static labels to nodes(only be set on node start), used to classify nodes. Labels(key, value) = Map<String, String>
+2. Improve the expression of NodeAffinity to support affinity\anti-affinity and labels expressions.
+3. The actual principle of NodeAffinitySchedulingStratgy is. According to the label matching expression in the strategy, traverse and search for LabelsResource of all nodes. Pick out nodes that meet the expression requirements.
+
+
+![NodeAffinity Concept](https://user-images.githubusercontent.com/11072802/217133358-19ddc916-bf15-4f69-96e1-9cbd157b0e1d.png)
+
+Scheduling Policy | Label Owner | Label select operator
+-- | -- | --
+NodeAffinity | Node | in, not_in, exists, does_not_exist
+
+### API Design
+
+**The apis of actors/nodes add labels**  
+This interface is already very simple, so we will not set up multiple solutions for everyone to discuss.
+```python
+#  Node add static labels. 
+ray start ... --labels={"location": "dc_1"}
+
+# Dynamically updating node labels is not implemented now, and will be considered in the second stage.
+```
+
+**The apis of the actor-affinity/node-affinity scheduling.**  
+
+**Option 1: Simplify through syntactic sugar**
+
+
+```python
+# Scheduled to a node with a specific IP.
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_in(key="node_ip", values=["xxx.xxx.xx.xx"], is_soft=false))
```

Comment:
Minor, but seems like we don't really need the `node_affinity` wrapper here?

Reply:
The benefits of wrapping with node_affinity here are as follows:
1. The label representation in node_affinity can be multiple. So it needs to be wrapped with node_affinity
```
# Multiple label expressions can be filled in at the same time, and the relationship between the expressions is "and". The dispatch must satisfy each expression.
# The actual meaning of this expression is that it must be scheduled to a node with a GPU, and as much as possible to a node with a GPU of the A100 type.
actor_1 = Actor.options(
        scheduling_strategy=node_affinity([
            label_in("gpu_type", ["A100"], true),
            label_exists("gpu_type", false)
        ])
    ).remote()
```
2. After actor_affinity also intends to introduce label expressions. At that time, actor_affinity & node_affinity can reuse the four expressions label_in\label_exist at the same time. The code implementation can be more concise and easier for users to understand.
```
Actor.options(
        scheduling_strategy=node_affinity(label_in(key="node_ip", values=["xxx.xxx.xx.xx"], is_soft=false)).remote()

Actor.options(
        scheduling_strategy=actor_affinity(label_in(key="actor_type", values=["dog"], is_soft=false)).remote()

```